### PR TITLE
single-operand cond

### PR DESCRIPTION
**Pull request by froystig**
This addresses the fan-out problem from #2775, and makes partial progress towards n-way conditionals per #2725.

### CODE REVIEW

File: jax/lax/lax_control_flow.py [hunk 1]
```diff
@@ -55,19 +57,59 @@
...
+
+  newvar = core.gensym('_')     # TODO(frostig): safer gensym
+  all_const_avals = tuple(
+      tuple(raise_to_shaped(core.get_aval(c)) for c in consts)
+      for consts in all_consts)
+  unused_const_vars = tuple(
+      tuple(newvar(aval) for aval in const_avals)
+      for const_avals in all_const_avals)
+
+  def pad_jaxpr_constvars(i, jaxpr):
```
Comment (gnecula) [inline jax/lax/lax_control_flow.py:92]: I would find some comments useful here

File: jax/lax/lax_control_flow.py [hunk 2]
```diff
@@ -55,19 +57,59 @@
...
   return typed_jaxpr, consts, out_tree()
 
+# TODO(frostig): Cache, as in _initial_style_jaxpr, or their common subroutine
+def _initial_style_jaxprs_with_common_consts(funs: Sequence[Callable],
+                                             in_tree, in_avals):
+  jaxprs, all_out_pvals, all_consts, all_out_trees = unzip4([
+      _initial_style_untyped_jaxpr(fun, in_tree, in_avals) for fun in funs])
+
+  newvar = core.gensym('_')     # TODO(frostig): safer gensym
+  all_const_avals = tuple(
```
Comment (gnecula) [inline jax/lax/lax_control_flow.py:85]: Isn't there room for optimization here? If both branches close over the same value, we should take it as input only once, to reduce the wasted space for the cotangents.

File: jax/lax/lax_control_flow.py [hunk 3]
```diff
@@ -744,6 +826,36 @@ def f_aug(*args):
 
   return _make_typed_jaxpr(f_aug, jaxpr.in_avals)
 
+def _join_cond_pe_staged_jaxpr_inputs(jaxprs, res_avals_per_jaxpr):
+  newvar = core.gensym('~')     # TODO(frostig): safer gensym
```
Comment (gnecula) [inline jax/lax/lax_control_flow.py:830]: Maybe you can pass the newvar in from the caller.

File: jax/lax/lax_control_flow.py [hunk 4]
```diff
@@ -744,6 +826,36 @@ def f_aug(*args):
 
   return _make_typed_jaxpr(f_aug, jaxpr.in_avals)
 
+def _join_cond_pe_staged_jaxpr_inputs(jaxprs, res_avals_per_jaxpr):
```
Comment (gnecula) [inline jax/lax/lax_control_flow.py:829]: I would find some comment helpful here.

File: jax/lax/lax_control_flow.py [hunk 5]
```diff
@@ -533,29 +584,63 @@ def cond(pred, true_operand, true_fun, false_operand, false_fun):
...
-  true_ops, true_tree = tree_flatten((true_operand,))
-  true_avals = tuple(_map(_abstractify, true_ops))
-  true_jaxpr, true_consts, true_out_tree = _initial_style_jaxpr(true_fun, true_tree, true_avals)
-  false_ops, false_tree = tree_flatten((false_operand,))
-  false_avals = tuple(_map(_abstractify, false_ops))
-  false_jaxpr, false_consts, false_out_tree = _initial_style_jaxpr(false_fun, false_tree, false_avals)
+  ops, ops_tree = tree_flatten((operand,))
+  ops_avals = tuple(_map(_abstractify, ops))
+
+  jaxprs, consts, out_trees = _initial_style_jaxprs_with_common_consts(
```
[the remark this answers is not in the document]
Comment (froystig) [reply jax/lax/lax_control_flow.py:590]: I structured it this way with n-way conditionals (#2725) in mind.

File: jax/lax/lax_control_flow.py (hunk 2 above)
Comment (froystig) [reply jax/lax/lax_control_flow.py:83]: Good point. The irrelevant arguments are unused/dropped within each jaxpr, but there may still be costs. Regarding cotangents, I also wonder if there is a more broadly-applicable optimization to make in our AD, if it's not in place already?

File: jax/lax/lax_control_flow.py (hunk 3 above)
Comment (froystig) [reply jax/lax/lax_control_flow.py:817]: I think we should fix the root issue and make a generally safer gensym (given a jaxpr), then use it here. The TODO is meant as a note to that end.

Comment (froystig) [reply jax/lax/lax_control_flow.py:841]: Resolved by #3211

**Pull request merged**